accepted/single-file/design.md

```diff
@@ -66,225 +57,209 @@ Publishing to a single file can be triggered by adding the following property to
+On Startup, the [host components](https://github.com/dotnet/core-setup/blob/master/Documentation/design-docs/host-components.md) perform the following functions:
 
-### New API
+* **AppHost**: The AppHost identifies itself as a single-file bundle (by checking the [bundle marker](bundler.md#bundle-marker)) before invoking [HostFxr](https://github.com/dotnet/core-setup/blob/master/Documentation/design-docs/host-components.md#host-fxr).
 
-In this section, we propose adding a few APIs to facilitate common operations on bundled-apps.
+* **HostFxr**: If invoked from a single-file app, HostFxr process the `runtimeconfig.json` and `deps.json` files directly from the bundle. The location of these `json` files  are identified directly from the bundle-header for simple access.
 
-The binaries that are published in the project are expected to be handled transparently by the host. However, explicit access to the embedded files is useful in situations such as:
+* **HostPolicy**: Much of the bundle-processing is performed within HostPolicy.
+  
+  * If the app needs extraction, extracts out the appropriate files as explained in [this document](extractor.md).
+  
+  * Reads the `deps.json` file directly from the bundle and resolves dependencies.
+  
+  * Processes the Bundle manifest to maintain internal data-structures to locate bundled assemblies when probed by the runtime.
+  
+  * Implements a bundle-probe function, which will be used by the runtime to probe contents of the bundle when resolving assemblies. 
+  
+    ```C++
+    /// <summary>
+    /// <param name="path"> Relative-path to the file being probed. </param>
+    /// <param name="size"> Out-param: size of the file, if found. </param>
+    /// <param name="path"> Out-param: offset within the bundle, if found</param>
+    /// <returns> true if the requested file is found in the bundle, 
+    ///           false otherwise. </returns>
+    /// </summary>
+    bool bundle_probe(const char *path, int64_t *size, int64_t *offset);
+    ```
+    
+  * The `bundle_probe` function pointer is passed to the runtime (encoded as a string) through a property named `BUNDLE_PROBE`.
+  
+  * When probing for assemblies, the [host probing logic](https://github.com/dotnet/runtime/blob/master/docs/design/features/host-probing.md#probing-paths) will treat bundled assemblies similar to assemblies in the app directory. The probe ordering will be in the order:  servicing location, the *single-file bundle*, app directory, framework directory(s) from higher to lower, shared store, additional specified probing paths.
+  
+  * The assemblies on disk are listed in `TRUSTED_PLATFORM_ASSEMBLIES` using absolute paths. As a distinguishing convention, the assemblies to be loaded from the single-file bundle will be listed in `TRUSTED_PLATFORM_ASSEMBLIES` using relative paths (as noted in the [bundle meta-data](bundler.md#Bundle-Transformation)). 
```

Comment:
The alternative option would be to just use the probing function, and not include embedded assemblies at all.

Why is it a good idea to include the embedded assemblies on `TRUSTED_PLATFORM_ASSEMBLIES`? Do we expect that it will make more things "work"?

Note that `TRUSTED_PLATFORM_ASSEMBLIES` list is used by code out there, and the code generally expect the paths to be actual file paths. One example from many: https://github.com/genesisdotnet/genesis/blob/master/src/Genesis/GenesisGlobals.cs + https://github.com/genesisdotnet/genesis/blob/8cb94f1fe2242bd9bb6c7ebdee4178ad35d7de99/src/Executors/Input/Genesis.Input.SwaggerUrl/SwaggerUrlReader.cs#L282

Reply:
I don't have a good reason to go either way. Your argument about code expecting to find the assemblies in the TPA can also be used to say that not including them at all is a problem as well. It's kind of a toss up what's worse, having "Wrong" file paths, or not having the files at all. I guess it depends on the app what effect that will have.

Given that the bundle stores a directory tree with files in it, it's in theory possible for the runtime to need to know where to look for the assembly in question. That would have to be provided by the hostpolicy (as a result of processing .deps.json). That said, I can't think of a case where this would actually happen though - managed assemblies should basically always remain in the "root', except for satellites which follow their own layout.

I searched github a bit and using the TPA as source of file paths (which are actually used as file paths) is by far the most common case. I found dozens of cases like that and only one where TPA was used just to list assembly names. Also, Roslyn actually has code in it which uses TPA to get file paths: 
https://github.com/dotnet/roslyn/blob/e704ca635bd6de70a0250e34c4567c7a28fa9f6d/src/Scripting/Core/Hosting/Resolvers/RuntimeMetadataReferenceResolver.cs#L186

Given this I tend to agree that we should probably not list the bundle assemblies in TPA at all. It will still break these apps, but at least it will break them consistently. Returning relative paths (most likely just `Assembly.dll` without any directory) has a higher chance of accidentally working, but with the wrong file being loaded.

We should also track work to add detection of this pattern into the linker analyzer - basically `AppContext.GetData("TRUSTED_PLATOFRM_ASSEMBLIES")` should always be marked as potentially dangerous for single-file scenarios.

Reply:
In light of the discussion above, I was about to change the look up mechanism to:
•	The Host doesn’t include bundled assemblies in `TRUSTED_PLATFORM_ASSEMBLIES` 
•	The runtime probes `TRUSTED_PLATFORM_ASSEMBLIES`  first (so that servicing locations get priority), and then probes the bundle.

However, for satellite assemblies (assuming that we want to leave them servicible), the `PLATFORM_RESOURCE_ROOTS` are a set of directory paths.  So, if we probe `PLATFORM_RESOURCE_ROOTS` first, some files in those paths may unintentionally override the ones found in the bundle.

So, I've written the logic as:
• If a bundled assembly is overridden by a servicing location, exclude it from the `bundle_probe` listing
• Probe in the bundle first, followed by `TRUSTED_PLATFORM_ASSEMBLIES` / `PLATFORM_RESOURCE_ROOTS`  as applicable.